Fake: fix conv_transpose2d striding

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82846

The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic.

eellison committed:
Fake: fix conv_transpose2d striding

[ghstack-poisoned]

ezyang reviewed (approved):
input sniffing is cool, we do it in other places too.

eellison committed:
Update on "Fake: fix conv_transpose2d striding"


The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic. 

[ghstack-poisoned]

ngimel commented on torch/_meta_registrations.py:
```diff
@@ -307,10 +307,17 @@ def calc_conv_nd_return_shape(
+
+    def pick_memory_format(device_hint):
+        if device_hint == "cuda":
+            if is_channels_last(input_tensor) and is_channels_last(weight):
```
I see cuda producing channels-last if only weight is channels last
```
In [8]: m=m.cuda().to(memory_format=torch.channels_last)

In [9]: inp=inp.cuda()

In [10]: m(inp).stride()
Out[10]: (968, 1, 88, 8)

In [11]: inp.stride()
Out[11]: (144, 36, 6, 1)
```

eellison replied:
Yea, it's either, misread the `||` as an `&&`:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ConvUtils.h#L316

ngimel replied:
Which means that all 1x1 convs will happily return channels-last format :-(

eellison replied:
Which makes this PR correct at least :-)

ngimel commented on torch/_meta_registrations.py:
```diff
@@ -307,10 +307,17 @@ def calc_conv_nd_return_shape(
-            return torch.channels_last
-        elif input_tensor.is_contiguous(memory_format=torch.contiguous_format):
+    def is_channels_last(ten):
+        return ten.is_contiguous(memory_format=torch.channels_last)
```
This is too permissive, e.g. weight tensors with 1x1 kernel will be channels-last contiguous, but they won't be channels-last for the purposes of memory format propagation

ngimel reviewed (commented):
The right thing would be to expose `suggest_memory_format` to python because this is what's used internally to decide whether something is channels-last. But for now probably a TODO will suffice

eellison committed:
Update on "Fake: fix conv_transpose2d striding"


The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic. 

[ghstack-poisoned]

eellison committed:
Update on "Fake: fix conv_transpose2d striding"


The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic. 

[ghstack-poisoned]

eellison committed:
Update on "Fake: fix conv_transpose2d striding"


The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic. 

[ghstack-poisoned]

ezyang commented:
This might be the root cause of https://github.com/pytorch/torchdynamo/issues/1274 it would be good to land it

eellison commented:
Yup, that's the intention. But also setting up cross-ref testing now..

eellison commented:
@pytorchbot help

eellison commented:
@pytorchbot --help

eellison commented:
@pytorchbot -h merge

eellison commented:
@pytorchbot merge --green

eellison commented:
@pytorchbot merge
